### PR TITLE
Improve `storage_alias` and make `UnlockAndUnreserveAllFunds` independent of the pallet

### PR DESCRIPTION
**Pull request by bkchr**
This pull request is doing two things: 

1. It improve `storage_alias` by making it more generic over the prefix to use. Currently `storage_alias` tries to guess based on the generics if the user wants to use the pallet name or a fixed identifier as prefix. This pull requests changes the macro to take an additional attribute that tells it what prefix type the user wants. Besides that a third prefix type is added, a type that implements `Get<&'static str>` can also be provided as prefix. See the docs in the code for more information on how this works.

2. `pallet_democracy::UnlockAndUnreserveAllFunds` is made independent of the presence of the `pallet-democracy` pallet. This means that the migration can be used without the pallet still being used in the runtime. The user only needs to implement the `UnlockConfig` for some type that is then passed to `UnlockAndUnreserveAllFunds` to forward certain information that are required by the migration.

Besides this the pull request  also puts some tests in `frame-support` into their own file to clean up the `lib.rs`.

### CODE REVIEW

File: frame/support/src/tests/storage_alias.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+		assert_eq!(1, Value::<Prefix>::get().unwrap());
+
+		// The prefix is the one we declared above.
+		assert_eq!(&b"Hello"[..], Value::<Prefix>::module_prefix());
+	});
+}
+
+#[docify::export]
+#[test]
+fn storage_alias_guess() {
```
Comment (ggwpez) [inline frame/support/src/tests/storage_alias.rs:159]: Maybe one test that the backwards compatibility works?

**Review by ggwpez — APPROVED** [review]
Just some nits

File: frame/support/src/tests/storage_alias.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
+// This file is part of Substrate.
```
[the remark this answers is not in the document]
Comment (bkchr) [reply frame/support/src/tests/storage_alias.rs:1]: I see no need here.

**Review by gupnik — APPROVED** [review]

File: frame/support/procedural/src/storage_alias.rs [hunk 3]
```diff
@@ -142,13 +112,22 @@ mod storage_types {
 	syn::custom_keyword!(StorageNMap);
 }
 
+/// The types of prefixes the storage alias macro supports.
+mod prefix_types {
+	// Use the verbatim passed name as prefix.
+	syn::custom_keyword!(verbatim);
+	// The passed type is a pallet and its pallet name should be used as prefix.
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:119]:
```suggestion
	// The input type is a pallet and its pallet name should be used as the prefix.
```

File: frame/support/procedural/src/storage_alias.rs [hunk 4]
```diff
@@ -142,13 +112,22 @@ mod storage_types {
...
+	syn::custom_keyword!(pallet_name);
+	// The passed type implements `Get<'static str>` and this str should be used as prefix.
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:121]:
```suggestion
	// The input type implements `Get<'static str>` and this `str` should be used as the prefix.
```

File: frame/support/procedural/src/storage_alias.rs [hunk 5]
```diff
@@ -508,20 +454,49 @@ impl Parse for Input {
 	}
 }
 
+/// The type of the prefix the storage alias is using.
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:457]:
```suggestion
/// Defines which type of prefix the storage alias is using.
```

File: frame/support/procedural/src/storage_alias.rs [hunk 6]
```diff
@@ -508,20 +454,49 @@ impl Parse for Input {
...
+#[derive(Clone, Copy)]
+enum PrefixType {
+	/// Use the compatibility way to determine the prefix.
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:460]:
```suggestion
	/// An appropriate prefix will be determined automatically.
```

File: frame/support/procedural/src/storage_alias.rs [hunk 7]
```diff
@@ -508,20 +454,49 @@ impl Parse for Input {
...
+	///
+	/// If generics are passed, this is assumed to be a pallet and the pallet name should be used.
+	/// Otherwise use the verbatim passed name as prefix.
+	Compatibility,
+	/// Use the verbatim ident as prefix.
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:465]:
```suggestion
	/// The provided ident/name will be used as the prefix.
```

File: frame/support/procedural/src/storage_alias.rs [hunk 8]
```diff
@@ -508,20 +454,49 @@ impl Parse for Input {
...
+	Verbatim,
+	/// Use the given type that needs to implement `PalletInfoAccess` to determine the name. This
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:467]:
```suggestion
	/// The provided type will be used to determine the prefix. This type must
	/// implement `PalletInfoAccess` which specifies the proper name. This
```

File: frame/support/procedural/src/storage_alias.rs [hunk 9]
```diff
@@ -508,20 +454,49 @@ impl Parse for Input {
...
+	/// name is then used as prefix.
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:468]:
```suggestion
	/// name is then used as the prefix.
```

File: frame/support/procedural/src/storage_alias.rs [hunk 10]
```diff
@@ -508,20 +454,49 @@ impl Parse for Input {
...
+	PalletName,
+	/// Use the given type that needs to implement `Get<'static str>` to determine the prefix.
```
Comment (sam0x17) [inline frame/support/procedural/src/storage_alias.rs:470]:
```suggestion
	/// Uses the provided type implementing `Get<'static str>` to determine the prefix.
```

File: frame/support/src/lib.rs [hunk 11]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
-/// Generate a new type alias for [`storage::types::StorageValue`],
-/// [`storage::types::StorageMap`], [`storage::types::StorageDoubleMap`]
-/// and [`storage::types::StorageNMap`].
+/// Generate a [`#[pallet::storage]`](pallet_macros::storage) alias outside of a pallet.
 ///
-/// Useful for creating a *storage-like* struct for test and migrations.
+/// This storage alias works similar to the [`#[pallet::storage]`](pallet_macros::storage)
+/// attribute macro. It supports [`storage::types::StorageValue`],
+/// [`storage::types::StorageMap`], [`storage::types::StorageDoubleMap`] and
+/// [`storage::types::StorageNMap`]. The main difference to the normal
```
Comment (sam0x17) [inline frame/support/src/lib.rs:137]: This gets a bit verbose in the generated docs, I would recommend changing each link in this little section to be like e.g.:

```rust
/// [`StorageMap`](`storage::types::StorageMap`)
```

This way the reader just sees `StorageMap` but can easily figure out where that type is located by clicking it.

File: frame/support/src/lib.rs [hunk 12]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
 	const TYPE_ID: [u8; 4] = *b"modl";
 }
 
-/// Generate a new type alias for [`storage::types::StorageValue`],
-/// [`storage::types::StorageMap`], [`storage::types::StorageDoubleMap`]
-/// and [`storage::types::StorageNMap`].
+/// Generate a [`#[pallet::storage]`](pallet_macros::storage) alias outside of a pallet.
 ///
-/// Useful for creating a *storage-like* struct for test and migrations.
+/// This storage alias works similar to the [`#[pallet::storage]`](pallet_macros::storage)
```
Comment (sam0x17) [inline frame/support/src/lib.rs:134]:
```suggestion
/// This storage alias works similarly to the [`#[pallet::storage]`](pallet_macros::storage)
```

File: frame/support/src/lib.rs [hunk 13]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
-///
-/// # fn main() {}
-/// ```
+/// The flexibility around declaring the storage prefix makes this macro very useful for
+/// writing migrations etc.
+///
+/// # Examples
+///
+/// There are different ways to declare the `prefix` to use. The `prefix` type can either be
+/// declared explicetly by passing it to the macro as attribute or by letting the macro guess
```
Comment (sam0x17) [inline frame/support/src/lib.rs:149]:
```suggestion
/// declared explicetly by passing it to the macro as an attribute or by letting the macro guess
```

File: frame/support/src/lib.rs [hunk 14]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
+/// on what the `prefix` type is. The `prefix` is always passed as first generic argument to
```
Comment (sam0x17) [inline frame/support/src/lib.rs:150]:
```suggestion
/// on what the `prefix` type is. The `prefix` is always passed as the first generic argument to
```

File: frame/support/src/lib.rs [hunk 15]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
+/// the type declaration. When using [`#[pallet::storage]`](pallet_macros::storage) this first
+/// generic argument is always `_`. Besides declaring the `prefix`, the rest of the type
+/// declaration works as with [`#[pallet::storage]`](pallet_macros::storage).
+///
+/// 1. Use the `verbatim` prefix type. This prefix type uses the given identifier as `prefix`:
```
Comment (sam0x17) [inline frame/support/src/lib.rs:155]:
```suggestion
/// 1. Use the `verbatim` prefix type. This prefix type uses the given identifier as the `prefix`:
```

File: frame/support/src/lib.rs [hunk 16]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
+#[doc = docify::embed!("src/tests/storage_alias.rs", verbatim_attribute)]
+///
+/// 2. Use the `pallet_name` prefix type. This prefix type uses name of pallet as configured in
```
Comment (sam0x17) [inline frame/support/src/lib.rs:158]:
```suggestion
/// 2. Use the `pallet_name` prefix type. This prefix type uses the name of the pallet as configured in
```

File: frame/support/src/lib.rs [hunk 17]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
+///    [`construct_runtime!`] as `prefix`:
```
Comment (sam0x17) [inline frame/support/src/lib.rs:159]:
```suggestion
///    [`construct_runtime!`] as the `prefix`:
```

File: frame/support/src/lib.rs [hunk 18]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
+#[doc = docify::embed!("src/tests/storage_alias.rs", pallet_name_attribute)]
+/// It requires that the given prefix type implements
+/// [`PalletInfoAccess`](traits::PalletInfoAccess) (which is always the case for FRAME pallet
+/// structs). In the example above `Pallet<T>` is the prefix type.
```
Comment (sam0x17) [inline frame/support/src/lib.rs:163]:
```suggestion
/// structs). In the example above, `Pallet<T>` is the prefix type.
```

File: frame/support/src/lib.rs [hunk 19]
```diff
@@ -127,80 +129,48 @@ impl TypeId for PalletId {
...
+///
+/// 3. Use the `dynamic` prefix type. This prefix type calls [`Get::get()`](traits::Get::get)
+///    to get the `prefix`:
+#[doc = docify::embed!("src/tests/storage_alias.rs", dynamic_attribute)]
+/// It requires that the given prefix type implements [`Get<'static str>`](traits::Get).
+///
+/// 4. Let the macro "guess" what kind of prefix type to use. This only supports verbatim or
+///    pallet name. The macro uses the presence of generic arguments to the prefix type as
+///    indication to use the pallet name:
```
Comment (sam0x17) [inline frame/support/src/lib.rs:172]:
```suggestion
///    an indication that it should use the pallet name as the `prefix`:
```

**Review by sam0x17 — APPROVED** [review]
found a bunch of minor grammar/wording nits. Might need to re-wrap a few of the doc paragraphs if you accept them

File: frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs [hunk 20]
```diff
@@ -18,15 +18,66 @@
...
+
+/// The configuration for [`UnlockAndUnreserveAllFunds`].
+pub trait UnlockConfig: 'static {
+	/// The account id as being used in the runtime.
+	type AccountId: Parameter + Ord;
+	/// The currency type being used in the runtime.
+	///
+	/// Should be the same as previously used for the pallet.
+	type Currency: LockableCurrency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+	/// The name of the pallet as previously configured in `construct_runtime!`.
```
Comment (bkchr) [inline frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs:47]:
```suggestion
	/// The name of the pallet as previously configured in [`construct_runtime!`].
```

File: frame/support/src/lib.rs (hunk 11 above)
Comment (bkchr) [inline frame/support/src/lib.rs:137]:
```suggestion
/// attribute macro. It supports [`StorageValue`](storage::types::StorageValue),
/// [`StorageMap`](storage::types::StorageMap), [`StorageDoubleMap`](storage::types::StorageDoubleMap) and
/// [`StorageNMap`](storage::types::StorageNMap). The main difference to the normal
```

**Comment by bkchr** [comment]
bot fmt

**Comment by bkchr** [comment]
bot merge

**Review by liamaharon — APPROVED** [review]
Perfect, thank you! 

I have implemented and tested this on my [Gov V1 unlock PR](https://github.com/paritytech/polkadot/pull/7314) (commit with changes: 
https://github.com/paritytech/polkadot/pull/7314/commits/5e440c161ad860de5bef74bd7ea6cf05b968454c) and it works great.

Additional benefit: we can wipe the pallet storage in the same migration we unlock the funds.

**Review by wentelteefje — APPROVED** [review]

**Pull request merged**